src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -304,29 +534,253 @@ private void OnMvcBeforeAction(object arg)
+                    }
+                }
+
+                // mirror the parent if we couldn't extract a route
+                span.ResourceName = resourceName ?? parentSpan.ResourceName;
+
+                mvcSpanTags.AspNetCoreAction = actionName;
+                mvcSpanTags.AspNetCoreController = controllerName;
+                mvcSpanTags.AspNetCoreArea = areaName;
+                mvcSpanTags.AspNetCorePage = pagePath;
+                mvcSpanTags.AspNetCoreRoute = aspNetRoute;
+
+                if (!isUsingEndpointRouting && isFirstExecution)
+                {
+                    // If we're using endpoint routing or this is a pipeline re-execution,
+                    // these will already be set correctly
+                    if (parentSpan.Tags is AspNetCoreTags parentTags)
+                    {
+                        parentTags.AspNetCoreRoute = aspNetRoute;
+                    }
+
+                    parentSpan.ResourceName = span.ResourceName;
+                }
+            }
+        }
+
+        private void OnRoutingEndpointMatched(object arg)
+        {
+            var tracer = _tracer ?? Tracer.Instance;
+
+            if (!tracer.Settings.IsIntegrationEnabled(IntegrationId) ||
+                !tracer.Settings.RouteTemplateResourceNamesEnabled)
+            {
+                return;
+            }
+
+            Span span = tracer.ActiveScope?.Span;
+
+            if (span != null)
+            {
+                var tags = span.Tags as AspNetCoreTags;
+                if (tags is null || !arg.TryDuckCast<HttpRequestInEndpointMatchedStruct>(out var typedArg))
+                {
+                    // Shouldn't happen in normal execution
+                    return;
+                }
+
+                HttpContext httpContext = typedArg.HttpContext;
+                var trackingFeature = httpContext.Features.Get<RequestTrackingFeature>();
+                var isFirstExecution = trackingFeature.IsFirstPipelineExecution;
+                if (isFirstExecution)
+                {
+                    trackingFeature.IsUsingEndpointRouting = true;
+                    trackingFeature.IsFirstPipelineExecution = false;
+
+                    var url = GetUrl(httpContext.Request);
+                    if (!string.Equals(url, trackingFeature.Url))
+                    {
+                        // URL has changed from original, so treat this execution as a "subsequent" request
+                        // Typically occurs for 404s for example
+                        isFirstExecution = false;
+                    }
+                }
+
+                // NOTE: This event is when the routing middleware selects an endpoint. Additional middleware (e.g
+                //       Authorization/CORS) may still run, and the endpoint itself has not started executing.
+
+                var rawEndpointFeature = httpContext.Features[EndpointFeatureType];
```

Comment:
Is there a scenario in which `EndpointFeatureType` is null?

Reply:
Good point - it _shouldn't_ be if we're in here, but better to be safe 👍 done in https://github.com/DataDog/dd-trace-dotnet/pull/1289/commits/75a6e197db80f988656d972bab82c992e318aac5